debian/patches/gflag-time-limit.diff

```diff
@@ -0,0 +1,139 @@
+ static time_t Gflag_time;		/* The last time_t the dump file was rotated. */
+@@ -278,8 +280,10 @@
+ #ifdef _WIN32
+     static HANDLE timer_handle = INVALID_HANDLE_VALUE;
+     static void CALLBACK verbose_stats_dump(PVOID param, BOOLEAN timer_fired);
++    static void CALLBACK gflag_timer_fired(PVOID param, BOOLEAN timer_fired);
+ #else /* _WIN32 */
+   static void verbose_stats_dump(int sig);
++  static void gflag_timer_fired(int sig);
+ #endif /* _WIN32 */
+ 
+ static void info(int);
+@@ -672,7 +676,7 @@
+ #define U_FLAG
+ #endif
+ 
+-#define SHORTOPTS "aAb" B_FLAG "c:C:d" D_FLAG "eE:fF:G:hHi:" I_FLAG j_FLAG J_FLAG "KlLm:M:nNOpq" Q_FLAG "r:s:StT:u" U_FLAG "vV:w:W:xXy:Yz:Z:#"
++#define SHORTOPTS "aAb" B_FLAG "c:C:d" D_FLAG "eE:fF:g:G:hHi:" I_FLAG j_FLAG J_FLAG "KlLm:M:nNOpq" Q_FLAG "r:s:StT:u" U_FLAG "vV:w:W:xXy:Yz:Z:#"
+ 
+ /*
+  * Long options.
+@@ -1631,13 +1635,27 @@
+ 			infile = optarg;
+ 			break;
+ 
++		case 'g':
++			gflag = atoi(optarg);
++			if (gflag < 1)
++				error("invalid number of seconds %s", optarg);
++
++			/* Grab the current time for time limit. */
++			if ((gflag_time = time(NULL)) == (time_t)-1) {
++				error("%s: can't get current time: %s",
++				    __func__, pcap_strerror(errno));
++			}
++
++			gflag_time += gflag;
++			break;
++
+ 		case 'G':
+ 			Gflag = atoi(optarg);
+ 			if (Gflag < 0)
+ 				error("invalid number of seconds %s", optarg);
+ 
+-                        /* We will create one file initially. */
+-                        Gflag_count = 0;
++			/* We will create one file initially. */
++			Gflag_count = 0;
+ 
+ 			/* Grab the current time for rotation use. */
+ 			if ((Gflag_time = time(NULL)) == (time_t)-1) {
+@@ -2509,6 +2527,16 @@
+ 		timer.it_value.tv_usec = 1;
+ 		setitimer(ITIMER_REAL, &timer, NULL);
+ #endif /* _WIN32 */
++	} else if (gflag > 0) {
++#ifdef _WIN32
++		CreateTimerQueueTimer(&timer_handle, NULL,
++		    gflag_timer_fired, NULL, gflag_time * 1000, 0,
++		    WT_EXECUTEDEFAULT|WT_EXECUTELONGFUNCTION);
++		setvbuf(stderr, NULL, _IONBF, 0);
++#else /* _WIN32 */
++		(void)setsignal(SIGALRM, gflag_timer_fired);
++		alarm(gflag_time);
++#endif /* _WIN32 */
+ 	}
+ 
+ 	if (RFileName == NULL) {
+@@ -3132,17 +3160,37 @@
+ 		pcap_dump_flush(pdd);
+ }
+ #endif
++
++static void
++gflag_time_exceeded(void)
++{
++	(void)fprintf(stderr, "Time limit reached: %d seconds\n", gflag);
++	info(1);
++	exit_tcpdump(0);
++}
+ 
+ static void
+ print_packets_captured (void)
+ {
+ 	static u_int prev_packets_captured, first = 1;
++	time_t t;
+ 
+ 	if (infodelay == 0 && (first || packets_captured != prev_packets_captured)) {
+ 		fprintf(stderr, "Got %u\r", packets_captured);
+ 		first = 0;
+ 		prev_packets_captured = packets_captured;
+ 	}
++
++	/* Get the current time */
++	if ((t = time(NULL)) == (time_t)-1) {
++		error("%s: can't get current_time: %s",
++			__func__, pcap_strerror(errno));
++	}
++
++	/* If the time hits the expected end time, exit */
++	if (t >= gflag_time) {
++		gflag_time_exceeded();
++	}
```

Comment:
Why is this needed? The timer should fire and handle this already, no?

Reply:
This is if SIGALRM is already in use for the verbose logging mode.

Reply:
Er. Right. But that's not gonna work if no gflag is specified? (gflag_time will be zero in that case, as it's statically allocated)

Reply:
Thanks for spotting the lack of the `if gflag` test - fixed.

The timer has two very different paths:
- Every second interval timer for printing stats
- Single timer for the gflag value

Both end up in `gflag_time_exceeded()` to actually do the exit.

But I don't think we can get much other code sharing here unless we handled the second case by an interval timer as well. Which didn't seem efficient; if we don't need any other periodic call then going straight to a single `alarm()` call seemed the most efficient way of handling it. The time checking only comes into play when SIGALARM is coopted by the interval timer.